### PR TITLE
cert-manager: fast-forward to upstream 3945595b

### PR DESCRIPTION
**Pull request by munnerz**
**What this PR does / why we need it**:

**This PR should not be merged yet**

This is a draft PR to update the Helm chart for the v0.5 cert-manager release.

I am opening it now ahead of time, as it involves a few larger changes than usual and I'd like to get review feedback ahead of cutting a release to ensure the PR can be accepted in its current form.

A breakdown of what is is included and corresponding PRs:

* added affinity and tolerations (jetstack/cert-manager#869)
* Add validating webhook and webhook tls autoconfiguration (jetstack/cert-manager#478)
* chart: annotate all CRDs with "crd-install" hook (jetstack/cert-manager#823)
* helm chart: remove endpoints from rbac resources (jetstack/cert-manager#769)

/hold
/cc @unguiculus

### CODE REVIEW

File: stable/cert-manager/Chart.yaml [hunk 1]
```diff
@@ -1,6 +1,6 @@
 name: cert-manager
-version: v0.4.1
-appVersion: v0.4.1
+version: v0.4.0-dev.6
+appVersion: v0.4.0-dev.0
```
Comment (munnerz) [inline stable/cert-manager/Chart.yaml]: These will both be bumped to v0.5

File: stable/cert-manager/README.md [hunk 2]
```diff
@@ -54,8 +54,8 @@ The following table lists the configurable parameters of the cert-manager chart
 | Parameter | Description | Default |
 | --------- | ----------- | ------- |
 | `image.repository` | Image repository | `quay.io/jetstack/cert-manager-controller` |
-| `image.tag` | Image tag | `v0.4.1` |
-| `image.pullPolicy` | Image pull policy | `IfNotPresent` |
+| `image.tag` | Image tag | `canary` |
+| `image.pullPolicy` | Image pull policy | `Always` |
```
Comment (munnerz) [inline stable/cert-manager/README.md]: This change will be reverted and updated to v0.5.0 accordingly

File: stable/cert-manager/templates/00-namespace.yaml [hunk 3]
```diff
@@ -3,4 +3,7 @@ apiVersion: v1
 kind: Namespace
 metadata:
   name: {{ .Release.Namespace | quote }}
+  labels:
+    name: {{ .Release.Namespace | quote }}
+    certmanager.k8s.io/disable-validation: "true"
```
Comment (munnerz) [inline stable/cert-manager/templates/00-namespace.yaml:8]: This file is only created when generating static deployment manifests, and the updates here are to keep it in line with default Helm 2.10 behaviour (and the second one is to help with the webhook deployment, and is effectively a no-op as we also match on the 'name' label)

File: stable/cert-manager/templates/certificate-crd.yaml [hunk 4]
```diff
@@ -3,6 +3,10 @@ apiVersion: apiextensions.k8s.io/v1beta1
 kind: CustomResourceDefinition
 metadata:
   name: certificates.certmanager.k8s.io
+{{- if semverCompare ">=2.10-0" .Capabilities.TillerVersion.SemVer }}
```
Comment (munnerz) [inline stable/cert-manager/templates/certificate-crd.yaml:6]: I noticed issues when adding the `helm.sh/hook: crd-install` hook with versions of Tiller that do not support the hook.

File: stable/cert-manager/templates/rbac.yaml [hunk 5]
```diff
@@ -13,12 +13,7 @@ rules:
     resources: ["certificates", "issuers", "clusterissuers"]
     verbs: ["*"]
   - apiGroups: [""]
-    # TODO: remove endpoints once 0.4 is released. We include it here in case
-    # users use the 'master' version of the Helm chart with a 0.2.x release of
-    # cert-manager that still performs leader election with Endpoint resources.
-    # We advise users don't do this, but some will anyway and this will reduce
-    # friction.
-    resources: ["endpoints", "configmaps", "secrets", "events", "services", "pods"]
+    resources: ["configmaps", "secrets", "events", "services", "pods"]
```
Comment (munnerz) [inline stable/cert-manager/templates/rbac.yaml:16]: This means we drop support for cert-manager v0.2.0-v0.2.2 (iirc). In general, our rule is v0.4.x of the chart is compatible with v0.4.y of cert-manager (i.e. we do not pin the patch version of appVersion and version, but we *do* pin minor versions).

File: stable/cert-manager/values.yaml [hunk 6]
```diff
@@ -5,8 +5,8 @@ replicaCount: 1
 
 image:
   repository: quay.io/jetstack/cert-manager-controller
-  tag: v0.4.1
-  pullPolicy: IfNotPresent
+  tag: canary
+  pullPolicy: Always
```
Comment (munnerz) [inline stable/cert-manager/values.yaml]: Will be updated to v0.5.0 and IfNotPresent

File: stable/cert-manager/values.yaml [hunk 7]
```diff
@@ -69,6 +69,9 @@ ingressShim: {}
   # defaultACMEChallengeType: ""
   # defaultACMEDNS01ChallengeProvider: ""
 
+webhook:
+  enabled: false
```
Comment (munnerz) [inline stable/cert-manager/values.yaml:73]: We disable the new webhook chart by default

**Review by munnerz — COMMENTED** [review]
Added a few comments to help with review!

**Comment by cpanato** [comment]
@munnerz please sign the DCO and

- since you are introducing some new things please bump the minor version or move to 1.0.0, since this chart is in the stable folder and was discussed in the helm chart meeting last week that all charts in stable should be at least 1.0.0 and not 0.x.y

side question: why use a dev image?

**Comment by munnerz** [comment]
Hey @cpanato,

re: DCO - will do. I am just putting a few changes together for this PR (i.e. the image tags). As I say, still WIP 😄 

> ...  in the helm chart meeting last week that all charts in stable should be at least 1.0.0 and not 0.x.y

This goes against our own versioning strategy - as explained here https://github.com/helm/charts/pull/7644#discussion_r216344802, to make user's lives easier, we pin our minor versions to each other so they can be sure that v0.4.x of the Helm chart works with (and deploys) v0.4.y of cert-manager. This makes users rolling back/forward far easier, and helps when users understand when they can and cannot use a simple `helm upgrade cert-manager ... --set image.tag=v0.4.y`.

We'd really like to not have to modify this behaviour, as I think it hurts usability and will generally just not help users at all.

As such, I'd like to request we can continue to use a v0.x.y format until cert-manager itself reaches (v1.0.0) (which will hopefully be within the next few months).

**Comment by munnerz** [comment]
I've updated the version numbers as mentioned in my above comments.

It'd be great if someone can verify the use of a subchart here is correct 😄

**Comment by munnerz** [comment]
Removed the WIP designation - if someone could take a look, I can then go ahead and release v0.5 on our end 😄

File: stable/cert-manager/webhook/Chart.yaml [hunk 8]
```diff
@@ -0,0 +1,5 @@
+apiVersion: v1
+appVersion: "v0.5.0"
+description: A Helm chart for Kubernetes
```
Comment (cpanato) [inline stable/cert-manager/webhook/Chart.yaml]: maybe add a better description of what that is.

File: stable/cert-manager/webhook/templates/ca-sync.yaml [hunk 9]
```diff
@@ -0,0 +1,172 @@
...
+      template:
+        metadata:
+          labels:
+            app: ca-helper
+        spec:
+          serviceAccountName: {{ include "webhook.fullname" . }}-ca-sync
+          restartPolicy: OnFailure
+          containers:
+          - name: ca-helper
+            image: quay.io/munnerz/apiextensions-ca-helper:canary
```
Comment (cpanato) [inline stable/cert-manager/webhook/templates/ca-sync.yaml]: Is this image correct?

File: stable/cert-manager/values.yaml (hunk 7 above)
Comment (cpanato) [reply stable/cert-manager/values.yaml:73]: can you add a test to validate when the webhook is true?

File: stable/cert-manager/webhook/templates/ca-sync.yaml (hunk 9 above)
Comment (munnerz) [reply stable/cert-manager/webhook/templates/ca-sync.yaml]: Updating to v0.1.0

**Comment by munnerz** [comment]
Addressed comments @cpanato

**Comment by cpanato** [comment]
@munnerz test failed because the image does not exist in quay

**Comment by munnerz** [comment]
I've now cut the v0.5.0 release, so docker images should be available now!

This should be good to go 😄

/retest
/hold cancel

**Comment by munnerz** [comment]
@mattfarina looks like the dco-labeler is not working - could you give it a kick? I am not too sure what events you have configured the labeler bot to listen for so not sure how to trigger it myself!

**Comment by cpanato** [comment]
/check-cla
//check-dco

**Comment by munnerz** [comment]
Thanks @cpanato! I will remember the incantation for next time 😄

**Comment by cpanato** [comment]
saw this error 
```
Error: customresourcedefinitions.apiextensions.k8s.io "certificates.certmanager.k8s.io" already exists
```

should we provide a test input to change that? how was it done in the past?

@mattfarina

**Comment by cpanato** [comment]
@munnerz maybe you can add a test values.yaml to set the `createCustomResource: true` to false

**Comment by munnerz** [comment]
I think that'd work in this instance, but is surely a temporary workaround?

I see two reasons for that error:

1) cert-manager is being used in the test cluster for some other purpose already, meaning the CRD exists (seems unlikely)
2) a previous `helm install` half succeeded (i.e. created the CRD resource), but eventually failed, meaning `helm delete` was not run (thus leaving the CRD installed).

Perhaps I am wrong 😄 setting `createCustomResource: false` would work, until that test cluster is rebuilt and the CRD is removed.

I've not run into this problem before, which leads me to think this is caused by (2) 😄

**Comment by cpanato** [comment]
lets try a retest again 😄 

/retest

**Comment by cpanato** [comment]
Then if the delete failed or did not run for any reason, we will need someone from the inside to make some cleanup

**Comment by mattfarina** [comment]
The issue is that the CRDs were not cleaned up, for some reason, in the test cluster on a previous run.

**Comment by Antiarchitect** [comment]
I have an issue that CRD's are not created at all.
Upgrading from 0.3.1 to 0.5.0 and having latest Tiller:
```
Client: &version.Version{SemVer:"v2.10.0", GitCommit:"9ad53aac42165a5fadc6c87be0dea6b115f93090", GitTreeState:"clean"}
Server: &version.Version{SemVer:"v2.10.0", GitCommit:"9ad53aac42165a5fadc6c87be0dea6b115f93090", GitTreeState:"clean"}
```
I have this while deploying:
```
Error: UPGRADE FAILED: unable to recognize "": no matches for kind "ClusterIssuer" in version "certmanager.k8s.io/v1alpha1"
```

**Comment by eldondevat** [comment]
@Antiarchitect are you still encountering this bug, I believe I am, too. Did you ever find a workaround?

**Pull request merged**